### PR TITLE
Rule S3693: Exception constructors should not throw exceptions

### LINKED ISSUE #584: Rule S3693: Exception constructors should not throw exceptions
[RSPEC-3693](https://jira.sonarsource.com/browse/RSPEC-3693)

It may be a good idea to raise an exception in a constructor if you're unable to fully flesh the object in question, but not in an exception constructor. If you do, you'll interfere with the exception that was originally being thrown. Further, it is highly unlikely that an exception raised in the creation of an exception will be properly handled in the calling code, and the unexpected, unhandled exception will lead to program termination.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #584

### CODE REVIEW

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExceptionConstructorShouldNotThrow.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        {
+            throw new FieldAccessException(); // Noncompliant
+        }
+    }
+
+    class MyException4 : Exception
+    {
+        public MyException4()
+        {
+            throw; // Noncompliant
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExceptionConstructorShouldNotThrow.cs]: is this a valid statement? what will be thrown here?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionConstructorShouldNotThrow.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class ExceptionConstructorShouldNotThrow : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3693";
+        private const string MessageFormat = "Avoid throwing exception in this constructor.";
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionConstructorShouldNotThrow.cs]: may be this should be `Avoid throwing exceptions in this constructor.` (e.g. plural exception)

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExceptionConstructorShouldNotThrow.cs (hunk 1 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/ExceptionConstructorShouldNotThrow.cs]: I don't think this is a valid statement in this condition but I guess the idea was to be sure the analyzer doesn't crash.

**Comment by Evangelink** [comment]
rebase done

**Pull request merged**